qa/rpc-tests/test_framework/mininode.py

```diff
@@ -359,57 +359,131 @@ def __repr__(self):
-    return ''.join(map(chr, vchCommitment))
+class CTxOutAsset(object):
+    def __init__(self, vchCommitment=b"\x00", vchRangeProof=b"", vchNonceCommitment=b""):
+        self.vchCommitment = vchCommitment
+        self.vchRangeProof = vchRangeProof
+        self.vchNonceCommitment = vchNonceCommitment
+
+    def deserialize(self, f):
+        version = ord(f.read(1))
+        if version == 0: self.vchCommitment = b'\x00'
+        elif version == 1: self.vchCommitment = b'\x01' + f.read(32)
+        elif version == 0xff: self.vchCommitment = b'\xff' + f.read(32)
+        elif version == 10 or version == 11: self.vchCommitment = bytes([version]) + f.read(32)
+        else: raise 'invalid CTxOutAsset in deserialize'
+
+    def serialize(self):
+        r = b""
+        r += self.vchCommitment
+        return r
+
+    def setToAsset(self, val):
+        if len(val) != 32:
+            raise 'invalid asset hash (expected 32 bytes got %d)' % len(val)
+        self.vchCommitment = b'\x01' + val
+
+    def __repr__(self):
+        return "CTxOutAsset(vchCommitment=%s vchRangeProof=%s, vchnonceCommitment=%s)" \
+            % (self.vchCommitment, self.vchRangeProof,
+               self.vchNonceCommitment)
 
-#Serialization only does value commitment
 class CTxOutValue(object):
-    def __init__(self, vchCommitment=[], vchRangeProof="", vchNonceCommitment=""):
+
+    def __init__(self, value=None):
+        self.setNull()
+        if value != None: self.setToAmount(value)
+
+    def setNull(self):
+        self.vchCommitment = b'\x00'
+        self.vchRangeProof =b''
+        self.vchNonceCommitment = b''
+
+    def deserialize(self, f):
+        version = ord(f.read(1))
+        if version == 0: self.vchCommitment = b'\x00'
+        elif version == 1: self.vchCommitment = b'\x01' + f.read(8)
+        elif version == 0xff: self.vchCommitment = b'\xff' + f.read(8)
+        elif version == 8 or version == 9: self.vchCommitment = bytes([version]) + f.read(32)
+        else: raise Exception('invalid CTxOutValue in deserialize. version %d' % version)
+
+    def serialize(self):
+        r = b""
+        if len(self.vchCommitment) < 1: raise ValueError('invalid commitment')
+        r += self.vchCommitment
+        return r
+
+    def setToAmount(self, amount):
+        commit = [1]*9
+        for i in range(8): #8 bytes
+            commit[8-i] = ((amount >> (i*8)) & 0xff)
+        self.vchCommitment = bytes(commit)
+
+    def getAmount(self):
+        if self.vchCommitment[0] != 1:
+            raise ValueError('getAmount() called on non-explicit CTxOutValue')
+        ret = 0
+        for i in range(8): #8 bytes
+            ret <<= 8
+            ret |= self.vchCommitment[i+1]
+        return ret
+
+    def __repr__(self):
+        return "CTxOutValue(vchCommitment=%s vchRangeProof=%s, vchnonceCommitment=%s)" \
+            % (self.vchCommitment, self.vchRangeProof,
+                self.vchNonceCommitment)
+
+class CTxOutNonce(object):
```

Comment:
Nonce class only has a 33-byte commitment, not a rangeproof or other witness data.